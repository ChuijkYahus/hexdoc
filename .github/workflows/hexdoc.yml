name: "[Reusable] Build and publish a hexdoc project"

on:
  workflow_call:
    inputs:
      python-version:
        description: Python version to install
        type: string
        required: true
      checkout:
        description: Whether to check out the branch specified in `branch`
        type: boolean
        required: true
      branch:
        description: Branch to generate docs from
        type: string
        required: false
      publish-latest-and-root:
        description: Overwrite `/v/latest` (and `/`, if releasing)
        type: boolean
        required: true
      publish-release:
        description: Overwrite `/v/<current_version>`
        type: boolean
        required: true
      bump-version-segment:
        description: Version segment to bump with Hatch
        type: string
        required: false
      props-path:
        description: Path to your props file (hexdoc.toml or properties.toml)
        type: string
        required: false
      pip-extras:
        description: Pip extras for your package (eg. [dev])
        type: string
        required: false
    outputs:
      publish-release:
        description: Value of publish-release input for convenience.
        value: ${{ inputs.publish-release }}
      pages-url:
        description: Current GitHub Pages url for this repo.
        value: ${{ jobs.build.outputs.pages-url }}

permissions:
  contents: read

env:
  UPDATE_LATEST: ${{ inputs.publish-latest-and-root }}
  RELEASE: ${{ inputs.publish-release }}

jobs:
  build:
    runs-on: ubuntu-latest

    env:
      GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}

    permissions:
      contents: write
      pages: read

    outputs:
      pages-url:  ${{ steps.get-url.outputs.pages-url }}
      matrix:     ${{ steps.list-langs.outputs.matrix }}
      branch:     ${{ steps.parse-inputs.outputs.branch }}
      props-path: ${{ steps.parse-inputs.outputs.props-path }}

    steps:
      - uses: actions/checkout@v3
      - uses: actions/setup-python@v4
        with:
          python-version: ${{ inputs.python-version }}
          cache: pip

      - name: Get Pages url
        id: get-url
        run: |
          url=$(gh api "repos/$GITHUB_REPOSITORY/pages" --jq '.html_url')
          echo "pages-url=$url" >> "$GITHUB_OUTPUT"
          echo "GITHUB_PAGES_URL=$url" >> "$GITHUB_ENV"

      - name: Parse inputs
        id: parse-inputs
        run: |
          if [[ ${{ inputs.checkout }} == true ]]; then
            branch='${{ inputs.branch }}'
            props_path=_checkout/${{ inputs.props-path }}
            echo "HATCH_GRADLE_DIR=_checkout" >> "$GITHUB_ENV"
          else
            branch=none
            props_path=${{ inputs.props-path }}
          fi

          echo "HEXDOC_PROPS=$props_path" >> "$GITHUB_ENV"

          echo "branch=$branch"         >> "$GITHUB_OUTPUT"
          echo "props-path=$props_path" >> "$GITHUB_OUTPUT"

      - name: Checkout input branch
        if: steps.parse-inputs.outputs.branch != 'none'
        uses: actions/checkout@v3
        with:
          ref: ${{ steps.parse-inputs.outputs.branch }}
          path: _checkout

      - name: Install docgen from source
        run: pip install .${{ inputs.pip-extras }} hatch

      - name: List book languages
        id: list-langs
        run: |
          echo "matrix=$(hexdoc list-langs)" >> "$GITHUB_OUTPUT"
          if [[ $RUNNER_DEBUG ]]; then
            tree -I '__pycache__|Common|Fabric|Forge|venv'
          fi

      - name: Export hexdoc resources
        run: hexdoc export

      - name: Bump version
        if: inputs.bump-version-segment
        run: hatch version "${{ inputs.bump-version-segment }}"

      - name: Commit changes
        uses: stefanzweifel/git-auto-commit-action@v4
        with:
          commit_message: Build web book from ${{ github.ref }}

      - name: Build package
        run: hatch build

      - name: Upload package artifact
        uses: actions/upload-artifact@v3
        with:
          name: hexdoc-build
          path: dist

      - name: Copy build to Pages
        if: env.UPDATE_LATEST
        run: |
          mkdir -p _site/v/latest
          cp -r dist _site/v/latest/dist

      - name: Upload temporary Pages artifact
        if: env.UPDATE_LATEST
        uses: actions/upload-artifact@v3
        with:
          name: github-pages-tmp
          path: _site

      - name: Add job summary
        run: echo "Built version \`$(hatch version)\` from commit \`$(git rev-parse --short "$GITHUB_SHA")\`." >> $GITHUB_STEP_SUMMARY

  generate:
    runs-on: ubuntu-latest
    needs: build
    continue-on-error: true

    strategy:
      fail-fast: false
      matrix:
        lang: ${{ fromJson(needs.build.outputs.matrix) }}

    env:
      GITHUB_PAGES_URL: ${{ needs.build.outputs.pages-url }}
      HEXDOC_PROPS: ${{ needs.build.outputs.props-path }}
      HEXDOC: hexdoc render _site --lang ${{ matrix.lang }} --clean

    steps:
      - uses: actions/checkout@v3
      - uses: actions/checkout@v3
        with:
          repository: object-Object/hexdoc
          path: _checkout_object-Object_hexdoc
      - uses: ./_checkout_object-Object_hexdoc/.github/actions/install-artifact-wheel
        with:
          name: hexdoc-build
          python-version: ${{ inputs.python-version }}
          pip-extras: ${{ inputs.pip-extras }}

      - name: Checkout input branch
        if: needs.build.outputs.branch != 'none'
        uses: actions/checkout@v3
        with:
          ref: ${{ needs.build.outputs.branch }}
          path: _checkout

      - name: Generate web book
        id: gen-normal
        continue-on-error: true
        run: $HEXDOC

      - name: Generate web book with missing translations
        if: steps.gen-normal.outcome == 'failure'
        run: $HEXDOC --allow-missing

      - name: Upload temporary Pages artifact
        uses: actions/upload-artifact@v3
        with:
          name: github-pages-tmp
          path: _site

      - name: Fail if the first generate step failed
        if: steps.gen-normal.outcome == 'failure'
        run: |
          echo "::error::Missing some i18n keys."
          exit 1

  deploy-pages:
    runs-on: ubuntu-latest
    needs: [build, generate]

    permissions:
      contents: write

    steps:
      - uses: actions/checkout@v3
      - uses: actions/checkout@v3
        with:
          repository: object-Object/hexdoc
          path: _checkout_object-Object_hexdoc
      - uses: ./_checkout_object-Object_hexdoc/.github/actions/install-artifact-wheel
        with:
          name: hexdoc-build
          python-version: ${{ inputs.python-version }}
          pip-extras: ${{ inputs.pip-extras }}

      - name: Checkout current Pages
        uses: actions/checkout@v3
        continue-on-error: true
        with:
          ref: gh-pages
          path: _site/dst

      - name: Download temporary Pages artifact
        uses: actions/download-artifact@v3
        with:
          name: github-pages-tmp
          path: _site/src/docs

      - name: Add new docs to site
        run: hexdoc merge --src _site/src/docs --dst _site/dst/docs

      - name: Deploy to Pages
        uses: JamesIves/github-pages-deploy-action@v4
        with:
          folder: _site/dst/docs
          target-folder: docs
