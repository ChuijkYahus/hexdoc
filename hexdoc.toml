modid = "hexdoc"
book = { "!None"={} }
default_lang = "en_us"

_export_root = "src/hexdoc/_export"
resource_dirs = [  # top takes priority
    { path="{^_export_root}/resources", reexport=false },
]
export_dir = "{_export_root}/generated"

template = { "!None"={} }

[minecraft_assets]
# https://github.com/PrismarineJS/minecraft-assets/tree/83e2169afbbce40990d69fc53e5962e4a793d467/data/1.19.1
ref = "83e2169afbbce40990d69fc53e5962e4a793d467"
version = "1.19.1"
