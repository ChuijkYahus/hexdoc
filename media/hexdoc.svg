<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<svg
   xmlns:dc="http://purl.org/dc/elements/1.1/"
   xmlns:cc="http://creativecommons.org/ns#"
   xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"
   xmlns:svg="http://www.w3.org/2000/svg"
   xmlns="http://www.w3.org/2000/svg"
   xmlns:sodipodi="http://sodipodi.sourceforge.net/DTD/sodipodi-0.dtd"
   xmlns:inkscape="http://www.inkscape.org/namespaces/inkscape"
   sodipodi:docname="hexdoc.svg"
   inkscape:version="1.0 (4035a4fb49, 2020-05-01)"
   inkscape:export-ydpi="48"
   inkscape:export-xdpi="48"
   inkscape:export-filename="C:\Users\jack\Pictures\hexdoc.png"
   id="svg8"
   version="1.1"
   viewBox="0 0 135.46666 135.46667"
   height="512"
   width="512">
  <defs
     id="defs2" />
  <sodipodi:namedview
     inkscape:window-maximized="1"
     inkscape:window-y="518"
     inkscape:window-x="-8"
     inkscape:window-height="1417"
     inkscape:window-width="2560"
     inkscape:snap-object-midpoints="false"
     inkscape:guide-bbox="true"
     showguides="true"
     units="px"
     showgrid="false"
     inkscape:document-rotation="0"
     inkscape:current-layer="layer1"
     inkscape:document-units="px"
     inkscape:cy="430.7964"
     inkscape:cx="5.0637413"
     inkscape:zoom="0.7"
     inkscape:pageshadow="2"
     inkscape:pageopacity="0.0"
     borderopacity="1.0"
     bordercolor="#666666"
     pagecolor="#ffffff"
     id="base">
    <sodipodi:guide
       id="guide837"
       orientation="0,512"
       position="0,135.46666" />
    <sodipodi:guide
       id="guide839"
       orientation="512,0"
       position="135.46666,135.46666" />
    <sodipodi:guide
       id="guide841"
       orientation="0,-512"
       position="135.46666,0" />
    <sodipodi:guide
       id="guide843"
       orientation="-512,0"
       position="0,0" />
    <sodipodi:guide
       inkscape:color="rgb(0,0,255)"
       inkscape:locked="false"
       inkscape:label=""
       id="guide847"
       orientation="-1,0"
       position="67.733332,67.733332" />
  </sodipodi:namedview>
  <metadata
     id="metadata5">
    <rdf:RDF>
      <cc:Work
         rdf:about="">
        <dc:format>image/svg+xml</dc:format>
        <dc:type
           rdf:resource="http://purl.org/dc/dcmitype/StillImage" />
        <dc:title></dc:title>
      </cc:Work>
    </rdf:RDF>
  </metadata>
  <g
     id="layer1"
     inkscape:groupmode="layer"
     inkscape:label="Layer 1">
    <path
       d="M 116.56666,95.94048 67.73333,124.14763 18.900005,95.94048 V 39.52618 L 67.73333,11.31903 116.56666,39.526181 Z"
       style="fill:#332233;fill-opacity:1;stroke:#201a20;stroke-width:16.9333;stroke-linecap:square;stroke-linejoin:round;stroke-miterlimit:4;stroke-dasharray:none;stroke-opacity:1;paint-order:stroke fill markers"
       id="path835" />
    <path
       id="path850"
       d="M 42.278415,67.733333 H 93.188249"
       style="fill:none;stroke:#dddddd;stroke-width:8.46667;stroke-linecap:round;stroke-linejoin:miter;stroke-miterlimit:4;stroke-dasharray:none;stroke-opacity:1" />
    <path
       style="fill:none;stroke:#dddddd;stroke-width:8.46667;stroke-linecap:round;stroke-linejoin:miter;stroke-miterlimit:4;stroke-dasharray:none;stroke-opacity:1"
       d="M 42.278416,87.771403 H 93.188249"
       id="path852" />
    <path
       id="path854"
       d="M 42.278416,47.695261 H 93.188249"
       style="fill:none;stroke:#dddddd;stroke-width:8.46667;stroke-linecap:round;stroke-linejoin:miter;stroke-miterlimit:4;stroke-dasharray:none;stroke-opacity:1" />
  </g>
</svg>
