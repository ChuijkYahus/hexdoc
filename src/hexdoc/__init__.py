"""
A [Jinja](https://jinja.palletsprojects.com/en/3.1.x/)-based documentation generator for [Patchouli](https://github.com/VazkiiMods/Patchouli) books.
"""

# attempt to reduce the amount of breaking changes from moving hexdoc out of hexmod
try:
    import hexdoc_hexcasting as hexcasting  # type: ignore
except ImportError:
    pass
