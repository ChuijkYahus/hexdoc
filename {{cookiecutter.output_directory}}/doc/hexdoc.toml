modid = "{{ cookiecutter.modid }}"
book = "{{ cookiecutter.book_id }}"
default_lang = "en_us"

resource_dirs = [  # top takes priority
    "resources",
    "{_common.src}/main/resources",
    "{_common.src}/generated/resources",
    {% if cookiecutter.multiloader -%}
    "{_fabric.src}/main/resources",
    "{_fabric.src}/generated/resources",
    "{_forge.src}/main/resources",
    "{_forge.src}/generated/resources",
    {%- endif %}
    { modid="hexcasting" },
    { modid="hexdoc" },
]
export_dir = "{{ cookiecutter.__export_root }}/generated"

[extra.hexcasting]
# regexes for parsing pattern registry files - try uncommenting a different one if your patterns aren't loading
# NOTE: "!Raw" means "don't apply variable interpolation to this value"

# Hex Casting (0.11.0)
{{ "# " if cookiecutter.pattern_regex != "hex_latest" }}_pattern_regex = { "!Raw"='make\(\s*"(?P<name>[a-zA-Z0-9_\/]+)",\s*(?:new )?(?:ActionRegistryEntry|OperationAction)\(\s*HexPattern\.fromAngles\(\s*"(?P<signature>[aqweds]+)",\s*HexDir.(?P<startdir>\w+)\)' }

# Hex Casting (0.10.3)
{{ "# " if cookiecutter.pattern_regex != "hex_0.10.3" }}_pattern_regex = { "!Raw"='HexPattern\.fromAngles\("(?P<signature>[qweasd]+)", HexDir\.(?P<startdir>\w+)\),\s*modLoc\("(?P<name>[^"]+)"\)[^;]+?(?:makeConstantOp|Op\w+|Widget\.\w+)(?:[^;]*(?P<is_per_world>true)\);)?' }

# Hexal (0.3.0)
{{ "# " if cookiecutter.pattern_regex != "hexal_0.3.0" }}_pattern_regex = { "!Raw"='make\(\s*"(?P<name>[a-zA-Z0-9_\/]+)",\s*HexPattern\.fromAngles\(\s*"(?P<signature>[aqweds]+)",\s*HexDir.(?P<startdir>\w+)\)' }

# Hexal (0.2.18)
{{ "# " if cookiecutter.pattern_regex != "hexal_0.2.18" }}_pattern_regex = { "!Raw"='(?s-m:HexPattern\.fromAngles\("(?P<signature>[qweasd]+)", HexDir\.(?P<startdir>\w+)\),\s*modLoc\("(?P<name>[^"]+)"\),[^,]+?(?:makeConstantOp|Op\w+).*?(?P<is_per_world>\btrue)?\)(?:[^\)]+?\bval\b|(?:(?!\bval\b)(?:.))+$))' }

[[extra.hexcasting.pattern_stubs]]
path = "{^^^_common.package}/{{ cookiecutter.pattern_registry }}"
regex = "{^_pattern_regex}"

[minecraft_assets]
# https://github.com/PrismarineJS/minecraft-assets/tree/83e2169afbbce40990d69fc53e5962e4a793d467/data/1.19.1
ref = "83e2169afbbce40990d69fc53e5962e4a793d467"
version = "1.19.1"

[template]
icon = "icon.png"
include = [
    "{{ cookiecutter.modid }}",
    "hexcasting",
    "patchouli",
    "hexdoc",
]

[template.args]
mod_name = "{{ cookiecutter.mod_display_name }}"
author = "{{ cookiecutter.author }}"
show_landing_text = false

# platforms

[_common]
src = "../{{ 'common/src' if cookiecutter.multiloader else 'src' }}"
package = "{src}/main/{{ cookiecutter.java_lang }}/{{ cookiecutter.java_package }}"

{% if cookiecutter.multiloader -%}
[_fabric]
src = "../fabric/src"
package = "{src}/main/{{ cookiecutter.java_lang }}/{{ cookiecutter.java_package }}/fabric"

[_forge]
src = "../forge/src"
package = "{src}/main/{{ cookiecutter.java_lang }}/{{ cookiecutter.java_package }}/forge"
{%- endif %}
