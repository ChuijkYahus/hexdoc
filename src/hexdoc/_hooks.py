from importlib.resources import Package
from pathlib import Path

import hexdoc
from hexdoc.plugin import (
    DefaultRenderedTemplatesImpl,
    HookReturn,
    LoadJinjaTemplatesImpl,
    LoadResourceDirsImpl,
    MinecraftVersionImpl,
    ModVersionImpl,
    hookimpl,
)


class HexdocPlugin(
    MinecraftVersionImpl,
    ModVersionImpl,
    LoadResourceDirsImpl,
    LoadJinjaTemplatesImpl,
    DefaultRenderedTemplatesImpl,
):
    @staticmethod
    @hookimpl
    def hexdoc_mod_version():
        return "(TODO: remove)"

    @staticmethod
    @hookimpl
    def hexdoc_minecraft_version() -> str:
        return "1.20.1"  # TODO: remove

    @staticmethod
    @hookimpl
    def hexdoc_load_resource_dirs() -> HookReturn[Package]:
        from hexdoc._export import generated, resources

        return [generated, resources]

    @staticmethod
    @hookimpl
    def hexdoc_load_jinja_templates() -> tuple[Package, str]:
        return hexdoc, "_templates"

    @staticmethod
    @hookimpl
    def hexdoc_default_rendered_templates(templates: dict[str | Path, str]) -> None:
        templates.update(
            {
                "index.html": "index.html.jinja",
                "index.css": "index.css.jinja",
                "textures.css": "textures.jcss.jinja",
                "index.js": "index.js.jinja",
            }
        )
