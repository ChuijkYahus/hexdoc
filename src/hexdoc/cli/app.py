import code
import json
import logging
import os
import shutil
import sys
from http.server import HTTPServer, SimpleHTTPRequestHandler
from pathlib import Path
from textwrap import dedent
from typing import Annotated, Union

from typer import Option, Typer

from hexdoc.core import ModResourceLoader
from hexdoc.minecraft import I18n, Tag
from hexdoc.minecraft.assets import (
    AnimatedTexture,
    HexdocAssetLoader,
    PNGTexture,
)
from hexdoc.utils.git import git_root

from . import render_block
from .utils.args import (
    DEFAULT_MERGE_DST,
    DEFAULT_MERGE_SRC,
    BranchOption,
    PathArgument,
    PropsOption,
    ReleaseOption,
    VerbosityOption,
)
from .utils.load import (
    load_book,
    load_common_data,
    render_textures_and_export_metadata,
)
from .utils.logging import repl_readfunc
from .utils.render import create_jinja_env, render_book
from .utils.sitemap import (
    assert_version_exists,
    delete_root_book,
    delete_updated_books,
    dump_sitemap,
    load_sitemap,
)

logger = logging.getLogger(__name__)

app = Typer(
    pretty_exceptions_enable=False,
    context_settings={
        "help_option_names": ["--help", "-h"],
    },
)
app.add_typer(render_block.app)


@app.command()
def list_langs(
    *,
    props_file: PropsOption,
    verbosity: VerbosityOption = 0,
):
    """Get the available language codes as a JSON list."""
    props, pm, _ = load_common_data(props_file, verbosity, "")
    with ModResourceLoader.load_all(
        props,
        pm,
        export=False,
    ) as loader:
        langs = sorted(I18n.list_all(loader))
        print(json.dumps(langs))


@app.command()
def repl(
    *,
    props_file: PropsOption,
    lang: Union[str, None] = None,
    allow_missing: bool = False,
    verbosity: VerbosityOption = 0,
):
    """Start a Python shell with some helpful extra locals added from hexdoc."""
    props, pm, _ = load_common_data(props_file, verbosity, "")

    loader = ModResourceLoader.load_all(
        props,
        pm,
        export=False,
    )

    _, book, i18n, context = load_book(
        props.book,
        props,
        pm,
        loader,
        lang,
        allow_missing,
    )

    repl_locals = dict(
        props=props,
        pm=pm,
        book=book,
        i18n=i18n,
        loader=loader,
        context=context,
    )

    code.interact(
        banner=dedent(
            f"""\
            [hexdoc repl] Python {sys.version}
            Locals: {', '.join(sorted(repl_locals.keys()))}"""
        ),
        readfunc=repl_readfunc(),
        local=repl_locals,
        exitmsg="",
    )


@app.command()
def export(
    output_dir: PathArgument = DEFAULT_MERGE_SRC,
    *,
    branch: BranchOption,
    release: ReleaseOption = False,
    props_file: PropsOption,
    allow_missing: bool = False,
    verbosity: VerbosityOption = 0,
):
    """Run hexdoc, but skip rendering the web book - just export the book resources."""
    props, pm, plugin = load_common_data(props_file, verbosity, branch)

    loader = ModResourceLoader.clean_and_load_all(
        props,
        pm,
        export=True,
    )

    site_path = plugin.site_path(versioned=release)
    asset_loader = plugin.asset_loader(
        loader=loader,
        site_url=f"{loader.props.url}/{site_path.as_posix()}",  # TODO: urlencode
        asset_url=props.env.asset_url,
        render_dir=output_dir / site_path,
    )

    all_metadata = render_textures_and_export_metadata(loader, asset_loader)

    i18n = I18n.load_all(
        loader,
        allow_missing=allow_missing,
    )[props.default_lang]

    if props.book:
        load_book(
            book_id=props.book,
            pm=pm,
            loader=loader,
            i18n=i18n,
            all_metadata=all_metadata,
        )


@app.command()
def render(
    output_dir: PathArgument = DEFAULT_MERGE_SRC,
    *,
    branch: BranchOption,
    release: ReleaseOption = False,
    clean: bool = False,
    lang: Union[str, None] = None,
    props_file: PropsOption,
    allow_missing: bool = False,
    verbosity: VerbosityOption = 0,
):
    """Export resources and render the web book."""

    # load data
    props, pm, plugin = load_common_data(props_file, verbosity, branch, book=True)

    if not props.book:
        raise ValueError("Expected a value for props.book, got None")
    if not props.template:
        raise ValueError("Expected a value for props.template, got None")

    site_path = plugin.site_book_path(
        lang=lang or props.default_lang,
        versioned=release,
    )

    books, all_metadata = load_books(
        props.book,
        props,
        pm,
        lang,
        allow_missing,
    )

    # set up Jinja
    env = create_jinja_env(pm, props.template.include, props_file)

    template_names = (
        props.template.render
        if props.template.override_default_render
        else pm.default_rendered_templates(props.template.include)
    ) | props.template.extend_render

    templates = {
        Path(path): env.get_template(template_name)
        for path, template_name in template_names.items()
    }

    if not templates:
        raise RuntimeError(
            "No templates to render, check your props.template configuration "
            f"(in {props_file.as_posix()})"
        )

    if clean:
        shutil.rmtree(output_dir, ignore_errors=True)

    lang_names = {
        book_lang: i18n.localize_lang() for book_lang, (_, i18n, _) in books.items()
    }

    for book_lang, (book, i18n, context) in books.items():
        png_textures = PNGTexture.get_lookup(context.textures)
        animations = list(AnimatedTexture.get_lookup(context.textures).values())

        render_book(
            props=props,
            pm=pm,
            lang=book_lang,
            lang_names=lang_names,
            book=book,
            i18n=i18n,
            templates=templates,
            output_dir=output_dir,
            site_path=site_path,
            all_metadata=all_metadata,
            png_textures=png_textures,
            animations=animations,
            allow_missing=allow_missing,
            version=plugin.mod_version,
        )

    logger.info("Done.")


@app.command()
def merge(
    *,
    src: Path = DEFAULT_MERGE_SRC,
    dst: Path = DEFAULT_MERGE_DST,
    release: ReleaseOption = False,
):
    # ensure at least the default language was built successfully
    if update_latest:
        assert_version_exists(root=src, version="latest")

    # TODO: figure out how to do this with pluggy (we don't have the props file here)
    # if is_release:
    #     assert_version_exists(src, GRADLE_VERSION)

    dst.mkdir(parents=True, exist_ok=True)

    # remove any stale data that we're about to replace
    delete_updated_books(src=src, dst=dst)
    if update_latest and release:
        delete_root_book(root=dst)

    # do the merge
    shutil.copytree(src=src, dst=dst, dirs_exist_ok=True)

    # rebuild the sitemap
    sitemap = load_sitemap(dst)
    dump_sitemap(dst, sitemap)


@app.command()
def serve(
    *,
    props_file: PropsOption,
    port: int = 8000,
    src: Path = DEFAULT_MERGE_SRC,
    dst: Path = DEFAULT_MERGE_DST,
    branch: BranchOption,
    release: bool = True,  # generally want --release for development, looks nicer
    clean: bool = False,
    lang: Union[str, None] = None,
    allow_missing: bool = False,
    verbosity: VerbosityOption = 0,
):
    book_root = dst
    relative_root = book_root.resolve().relative_to(Path.cwd())

    base_url = f"http://localhost:{port}"
    book_url = f"{base_url}/{relative_root.as_posix()}"

    repo_root = git_root(props_file.parent)
    asset_root = repo_root.relative_to(Path.cwd())

    os.environ |= {
        # prepend a slash to the path so it can find the texture in the local repo
        # eg. http://localhost:8000/_site/src/docs/Common/...
        # vs. http://localhost:8000/Common/...
        "DEBUG_GITHUBUSERCONTENT": f"{base_url}/{asset_root.as_posix()}",
        "GITHUB_PAGES_URL": book_url,
    }

    print("Exporting...")
    export(
        branch=branch,
        release=release,
        props_file=props_file,
        allow_missing=allow_missing,
        verbosity=verbosity,
    )

    print("Rendering...")
    render(
        branch=branch,
        props_file=props_file,
        output_dir=src,
        release=release,
        clean=clean,
        lang=lang,
        allow_missing=allow_missing,
        verbosity=verbosity,
    )

    print("Merging...")
    merge(
        src=src,
        dst=dst,
        release=release,
    )

    print(f"Serving web book at {book_url} (press ctrl+c to exit)\n")
    with HTTPServer(("", port), SimpleHTTPRequestHandler) as httpd:
        # ignore KeyboardInterrupt to stop Typer from printing "Aborted."
        # because it keeps printing after nodemon exits and breaking the output
        try:
            httpd.serve_forever()
        except KeyboardInterrupt:
            pass


if __name__ == "__main__":
    app()
