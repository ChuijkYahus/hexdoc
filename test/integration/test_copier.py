# pyright: reportUnknownMemberType=false

import subprocess
import sys
from pathlib import Path

import pytest
from hexdoc.cli import ci
from pytest import MonkeyPatch
from syrupy.assertion import SnapshotAssertion

from ..conftest import list_directory
from ..tree import write_file_tree


def run_pip(*args: str):
    return subprocess.run((sys.executable, "-m", "pip") + args, check=True)


@pytest.fixture(scope="session")
def output_dir(monkeysession: MonkeyPatch, env_overrides: dict[str, str]):
    monkeysession.chdir("submodules/hexdoc-hexcasting-template/.ctt/test_copier")

    write_file_tree(
        ".",
        {
            "src/generated/resources": {},
            "src/main/java/com/package/Patterns.java": (
                """\
                public static final ActionRegistryEntry GET_CASTER = make(
                    "foobar",
                    new ActionRegistryEntry(
                        HexPattern.fromAngles("qaqaqaqaqaq", HexDir.NORTH_EAST),
                        OpGetCaster.INSTANCE
                    )
                );
                """
            ),
            "src/main/resources/assets": {
                "hexcasting/patchouli_books/thehexbook/en_us": {
                    "categories/foo.json": {
                        "name": "hexdoc.mod.title",
                        "icon": "minecraft:amethyst_shard",
                        "description": "hexcasting.category.basics.desc",
                        "sortnum": 0,
                    },
                    "entries/bar.json": {
                        "name": "hexdoc.welcome.header",
                        "category": "hexcasting:foo",
                        "icon": "minecraft:textures/mob_effect/nausea.png",
                        "sortnum": 0,
                        "advancement": "hexcasting:y_u_no_cast_angy",
                        "pages": [
                            {
                                "type": "patchouli:text",
                                "text": "bar.link.patterns",
                            },
                        ],
                    },
                    # expect the entry to be spoilered but the category to not be
                    "entries/baz.json": {
                        "name": "hexdoc.welcome.header",
                        "category": "hexcasting:basics",
                        "icon": "minecraft:textures/mob_effect/nausea.png",
                        "sortnum": 0,
                        "advancement": "hexcasting:enlightenment",
                        "pages": [
                            {
                                "type": "patchouli:text",
                                "text": "bar.link.patterns",
                            },
                        ],
                    },
                },
                "mod/lang/en_us.json": {
                    # make sure we're specifically testing internal and external links
                    "bar.link.patterns": "$(l:patterns)patterns$(l:foo)foo/$",
                },
            },
            ".env": "\n".join(
                f"{name}={value}" for name, value in env_overrides.items()
            ),
            "gradle.properties": (
                """\
                modVersion=1.0.0
                hexcastingVersion=0.11.1-7
                minecraftVersion=1.20.1
                """
            ),
        },
    )

    run_pip("install", ".", "--force-reinstall", "--no-deps")

    monkeysession.setenv("GITHUB_OUTPUT", "github_output.txt")
    monkeysession.setenv("GITHUB_REF_NAME", "main")
    monkeysession.setenv("MOCK_GITHUB_PAGES_URL", "GITHUB_PAGES_URL")

    ci.build(
        props_file=Path("doc/hexdoc.toml"),
        release=False,
    )

    yield Path("_site/src/docs").resolve()


@pytest.mark.copier
def test_index(output_dir: Path, path_snapshot: SnapshotAssertion):
    path_snapshot._custom_index = "vlatestmainen_usindex.html"  # pyright: ignore[reportPrivateUsage]
    assert output_dir / "v/latest/main/en_us/index.html" == path_snapshot


@pytest.mark.copier
def test_list_directory(output_dir: Path, snapshot: SnapshotAssertion):
    assert list_directory(output_dir) == snapshot
