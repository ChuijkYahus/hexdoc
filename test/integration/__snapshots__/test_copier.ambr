# serializer version: 1
# name: test_list_directory
  list([
    'v',
    'v/latest',
    'v/latest/main',
    'v/latest/main/dist',
    'v/latest/main/dist/hexdoc_mod-1.0.0.1.0.dev0-py3-none-any.whl',
    'v/latest/main/dist/hexdoc_mod-1.0.0.1.0.dev0.tar.gz',
    'v/latest/main/en_us',
    'v/latest/main/en_us/.sitemap-marker.json',
    'v/latest/main/en_us/apple-touch-icon-114x114.png',
    'v/latest/main/en_us/apple-touch-icon-120x120.png',
    'v/latest/main/en_us/apple-touch-icon-144x144.png',
    'v/latest/main/en_us/apple-touch-icon-152x152.png',
    'v/latest/main/en_us/apple-touch-icon-167x167.png',
    'v/latest/main/en_us/apple-touch-icon-180x180.png',
    'v/latest/main/en_us/apple-touch-icon-57x57.png',
    'v/latest/main/en_us/apple-touch-icon-60x60.png',
    'v/latest/main/en_us/apple-touch-icon-72x72.png',
    'v/latest/main/en_us/apple-touch-icon-76x76.png',
    'v/latest/main/en_us/bar',
    'v/latest/main/en_us/bar/index.html',
    'v/latest/main/en_us/basics',
    'v/latest/main/en_us/basics/index.html',
    'v/latest/main/en_us/baz',
    'v/latest/main/en_us/baz/index.html',
    'v/latest/main/en_us/casting',
    'v/latest/main/en_us/casting/index.html',
    'v/latest/main/en_us/favicon-16x16.png',
    'v/latest/main/en_us/favicon-180x180.png',
    'v/latest/main/en_us/favicon-196x196.png',
    'v/latest/main/en_us/favicon-32x32.png',
    'v/latest/main/en_us/favicon-64x64.png',
    'v/latest/main/en_us/favicon-96x96.png',
    'v/latest/main/en_us/favicon.ico',
    'v/latest/main/en_us/foo',
    'v/latest/main/en_us/foo/index.html',
    'v/latest/main/en_us/greatwork',
    'v/latest/main/en_us/greatwork/index.html',
    'v/latest/main/en_us/hexcasting.js',
    'v/latest/main/en_us/hexcasting_render.js',
    'v/latest/main/en_us/icon.png',
    'v/latest/main/en_us/index.css',
    'v/latest/main/en_us/index.html',
    'v/latest/main/en_us/index.js',
    'v/latest/main/en_us/interop',
    'v/latest/main/en_us/interop/index.html',
    'v/latest/main/en_us/items',
    'v/latest/main/en_us/items/index.html',
    'v/latest/main/en_us/lore',
    'v/latest/main/en_us/lore/index.html',
    'v/latest/main/en_us/mstile-270x270.png',
    'v/latest/main/en_us/mstile-310x150.png',
    'v/latest/main/en_us/mstile-310x310.png',
    'v/latest/main/en_us/mstile-70x70.png',
    'v/latest/main/en_us/patterns',
    'v/latest/main/en_us/patterns/great_spells',
    'v/latest/main/en_us/patterns/great_spells/index.html',
    'v/latest/main/en_us/patterns/index.html',
    'v/latest/main/en_us/patterns/spells',
    'v/latest/main/en_us/patterns/spells/index.html',
    'v/latest/main/en_us/textures.css',
  ])
# ---
